### PR TITLE
support huge amount of datacubes

### PR DESCRIPTION
**Pull request by bikov**
We have a large number of datasources in our use - and when trying to use Turnilo we are getting `{"error": "Can't fetch settings", "message": "Invalid string length"}` when querying `/sources`
This should help us to support our use case with a large number of datacubes by splitting datacubes API from clusters API and adding pagination to the `/sources/dataCubes`

### CODE REVIEW

**Comment by mkuthan** [comment]
Thank you for the contribution, we'll review the changes. Due to a Christmas break you could expect the review in the beginning of the new year.

File: src/client/utils/ajax/ajax.ts [hunk 1]
```diff
@@ -69,17 +70,26 @@ export class Ajax {
...
+  static async sources(appSettings: ClientAppSettings): Promise<ClientSources> {
+    try {
+      const headers = Ajax.headers(appSettings.oauth);
+      const clusters = (await axios.get("sources/clusters", { headers })).data;
+      let dataCubesResult: SerializedDataCube[] = [];
+      let isDone = false;
+      let batch = 0;
+      do {
+        const { dataCubes, isDone: isDoneResult } = (await axios.get(`sources/dataCubes?batch=${batch}`, { headers })).data;
+        dataCubesResult = dataCubesResult.concat(dataCubes);
```
Comment (shacharHagu) [inline src/client/utils/ajax/ajax.ts:86]: I prefer to use [... dataCubesResult,... dataCubes]

File: src/client/utils/ajax/ajax.ts [hunk 2]
```diff
@@ -69,17 +70,26 @@ export class Ajax {
...
-      })
-      .then(sourcesJS => deserialize(sourcesJS, appSettings));
+  static async sources(appSettings: ClientAppSettings): Promise<ClientSources> {
+    try {
+      const headers = Ajax.headers(appSettings.oauth);
+      const clusters = (await axios.get("sources/clusters", { headers })).data;
+      let dataCubesResult: SerializedDataCube[] = [];
+      let isDone = false;
+      let batch = 0;
+      do {
```
Comment (shacharHagu) [inline src/client/utils/ajax/ajax.ts:84]: What is the advantage of do.. while here?

File: src/client/utils/ajax/ajax.ts [hunk 3]
```diff
@@ -69,17 +70,26 @@ export class Ajax {
...
-    return axios.get<SerializedSources>("sources", { headers })
-      .then(resp => resp.data)
-      .catch(error => {
-        throw mapOauthError(appSettings.oauth, error);
-      })
-      .then(sourcesJS => deserialize(sourcesJS, appSettings));
+  static async sources(appSettings: ClientAppSettings): Promise<ClientSources> {
+    try {
+      const headers = Ajax.headers(appSettings.oauth);
+      const clusters = (await axios.get("sources/clusters", { headers })).data;
```
Comment (adrianmroz-allegro) [inline src/client/utils/ajax/ajax.ts:80]: Let's put those fetches (`fetchClusters` and `fetchDataCubes`) in separate async functions. Then we can run them in parallel instead of waiting between them.

File: src/client/utils/ajax/ajax.ts [hunk 4]
```diff
@@ -69,17 +70,26 @@ export class Ajax {
...
+      let dataCubesResult: SerializedDataCube[] = [];
+      let isDone = false;
+      let batch = 0;
```
Comment (adrianmroz-allegro) [inline src/client/utils/ajax/ajax.ts:83]: More like `page`? `batch` is a loaded word in this domain.

File: src/server/routes/sources/sources.ts [hunk 5]
```diff
@@ -16,33 +16,57 @@
...
+                error: "Can't fetch settings",
+                message: error.message
+            });
+        }
+    });
+    router.get("/dataCubes", async (req: Request, res: Response) => {
+        try {
+            const { dataCubes } = await settings.getSources();
+            const relevantDatasources = dataCubes.filter(dataCube => checkAccess(dataCube, req.headers))
+                .filter(dc => isQueryable(dc))
```
Comment (adrianmroz-allegro) [inline src/server/routes/sources/sources.ts:48]: Extract this `filter(isQueryable)` and `map(serializeDataCube)` to separate function in `models/sources/sources.ts`. Something like `serializeDataCubes`.

File: src/server/routes/sources/sources.ts [hunk 6]
```diff
@@ -16,33 +16,57 @@
...
+            res.status(500).send({
+                error: "Can't fetch settings",
+                message: error.message
+            });
+        }
+    });
+    router.get("/dataCubes", async (req: Request, res: Response) => {
+        try {
+            const { dataCubes } = await settings.getSources();
+            const relevantDatasources = dataCubes.filter(dataCube => checkAccess(dataCube, req.headers))
```
Comment (adrianmroz-allegro) [inline src/server/routes/sources/sources.ts:47]: Don't mix data cubes and data sources.

File: src/client/utils/ajax/ajax.ts [hunk 7]
```diff
@@ -69,17 +70,26 @@ export class Ajax {
...
+      while (!isDone) {
```
Comment (adrianmroz-allegro) [inline src/client/utils/ajax/ajax.ts:84]: I'm wondering if we could extract this to async iterator and with `for await ... of`. That's very loose idea though!

**Comment by bikov** [comment]
@adrianmroz-allegro Thank you for the review - I've fixed the comments

**Comment by adrianmroz-allegro** [comment]
@bikov We had long weekend there - sorry for delay. PR is here: https://github.com/allegro/turnilo/pull/1012 I'm still working on good tests for this, because we won't be using this feature so we'd like to have some confidence there.
